client/src/templates/Challenges/classic/show.tsx

```diff
@@ -301,6 +301,38 @@ function ShowClassic({
     setUsingKeyboardInTablist(usingKeyboardInTablist);
   };
 
+  // leave page confirmation if code is not saved
+
+  useEffect(() => {
+    let isFilesSaved = true;
+
+    const savedChallengeFiles = savedChallenges[0]?.challengeFiles;
+
+    if (!challengeFiles) return;
+
+    for (let index = 0; index < challengeFiles.length; index++) {
+      const hasChangedContentAfterSave =
+        savedChallengeFiles &&
+        savedChallengeFiles[index]?.contents != challengeFiles[index]?.contents;
+
+      if (!savedChallengeFiles || hasChangedContentAfterSave) {
+        isFilesSaved = false;
+      }
+    }
+
+    if (isFilesSaved) return;
+
+    function onbeforeUnload(e: BeforeUnloadEvent) {
+      e.preventDefault();
+    }
+
+    window.addEventListener('beforeunload', onbeforeUnload);
+
+    return () => {
+      window.removeEventListener('beforeunload', onbeforeUnload);
```

Comment:
```suggestion
    const savedChallengeFiles = savedChallenge?.challengeFiles;
    const savedContents = savedChallengeFiles?.map(file => file.contents);
    const currentContents = challengeFiles?.map(file => file.contents);

    const isSaved =
      savedChallengeFiles && isEqual(savedContents, currentContents);

    function handleBeforeUnload(e: BeforeUnloadEvent) {
      e.preventDefault();
    }

    if (!isSaved) window.addEventListener('beforeunload', handleBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', handleBeforeUnload);
```
This is a bit simpler and easier to read, I think. `isEqual` would need importing from `lodash-es`, though. `savedChallenge` would come from the selector I mentioned above.